### PR TITLE
Composite trainer

### PR DESCRIPTION
**Pull request by Solonets**
Hi once again and thank you for the great work. I was working on a project with your awesome lib and had a use case where I needed to optimize different parameters with different optimizers. Your yourself seem to see a point in this for example when training separately weights of the network with complex optimizers and primitive encodings (i.e frequency) with more trivial optimizers. You also seem to be applying different regularizers to different weights of the net, which you had to implement in many solvers. 

I went ahead and implemented a composite trainer which just runs several trainers, I hope it could be helpful.

### CODE REVIEW

**Comment by Solonets** [comment]
@Tom94 I see there is a problem with the custom_weight() method. But I think I can fix it if you think it is worth it.

File: include/tiny-cuda-nn/optimizers/composite.h [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    return {{"otype", "Composite"}, {"nested", nested}, {"learning_rate_scale", learning_rate_scale}};
+  }
+
+  json serialize() const override {
+    json::array_t nested;
+    for (auto& n : m_nested) {
+      nested.emplace_back(n->serialize());
+    }
+    return {{"nested", nested}};
```
Comment (Tom94) [inline include/tiny-cuda-nn/optimizers/composite.h:127]: Learning rate is missing here.

File: include/tiny-cuda-nn/optimizers/composite.h [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    const json::array_t& nested = params["nested"];
+    m_n_weights = 0;
+    for (size_t i = 0; i < nested.size(); ++i) {
+      m_offsets.emplace_back(m_n_weights);
+      m_n_weights += nested[i].value("n_params_to_optimize", 0);
+      m_nested.emplace_back(std::shared_ptr<Optimizer<T>>(create_optimizer<T>(nested[i])));
+    }
+    m_offsets.emplace_back(m_n_weights);
+    update_hyperparams(params);
```
Comment (Tom94) [inline include/tiny-cuda-nn/optimizers/composite.h:53]: Missing `set_learning_rate` call

File: include/tiny-cuda-nn/optimizers/composite.h [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+  void update_hyperparams(const json& params) override {
+    if (params.contains("nested") && params["nested"].is_array()) {
+      const json::array_t& nested = params["nested"];
+      for (int i = 0; i < m_nested.size(); i++) {
+        m_nested[i]->update_hyperparams(nested[i]);
+      }
+    }
+    if (params.contains("learning_rate_scale")) {
+      learning_rate_scale = params["learning_rate_scale"];
```
Comment (Tom94) [inline include/tiny-cuda-nn/optimizers/composite.h:103]: Use `set_learning_rate` instead

File: include/tiny-cuda-nn/optimizers/composite.h [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    return {{"nested", nested}};
+  }
+
+  void deserialize(const json& data) override {
+    const json::array_t& nested = data["nested"];
+    for (int i = 0; i < m_nested.size(); i++) {
+      m_nested[i]->deserialize(nested[i]);
+    }
+    learning_rate_scale = data["learning_rate_scale"];
```
Comment (Tom94) [inline include/tiny-cuda-nn/optimizers/composite.h:135]: Use `set_learning_rate` instead

**Review by Tom94 — CHANGES_REQUESTED** [review]
This is great, thank you! See my review in the comments.

File: README.md [hunk 5]
```diff
@@ -204,6 +204,7 @@ Following is a summary of the components of this framework. [The JSON documentat
 | EMA | `include/tiny-cuda-nn/optimizers/average.h` | Wraps another optimizer and computes an exponential moving average of the weights. The average is used for inference only (does not feed back into training).
 | Exponential Decay | `include/tiny-cuda-nn/optimizers/exponential_decay.h` | Wraps another optimizer and performs piecewise-constant exponential learning-rate decay.
 | Lookahead | `include/tiny-cuda-nn/optimizers/lookahead.h` | Wraps another optimizer, implementing the lookahead algorithm [[Zhang et al. 2019]](https://arxiv.org/abs/1907.08610).
+| Composite | `include/tiny-cuda-nn/optimizers/composite.h` | Allows using several optimizers on different parameters
```
[the remark this answers is not in the document]
Comment (Solonets) [reply README.md:207]: it wasn't ordered before, I would assume you have 3 most common on the top, and then the rest in the alphabetical order?

File: include/tiny-cuda-nn/optimizer.h [hunk 6]
```diff
@@ -55,9 +55,12 @@ class Optimizer : public ObjectWithMutableHyperparams {
 	virtual T* custom_weights() const = 0;
 
 	virtual bool supports_nesting() const = 0;
-	virtual const std::shared_ptr<Optimizer<T>>& nested() const {
+	virtual const std::shared_ptr<Optimizer<T>>& nested(uint32_t idx = 0) const {
 		throw std::runtime_error{"Optimizer does not support nesting."};
 	}
+	virtual uint32_t num_nesting() const {
```
[the remark this answers is not in the document]
Comment (Tom94) [reply include/tiny-cuda-nn/optimizer.h:61]: That's a great idea!

File: include/tiny-cuda-nn/optimizers/composite.h [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+#include <tiny-cuda-nn/optimizer.h>
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+TCNN_NAMESPACE_BEGIN
+
+std::vector<std::pair<uint32_t, uint32_t>> slice_weights(std::vector<std::pair<uint32_t, uint32_t>> object_layer_size,
```
[the remark this answers is not in the document]
Comment (Solonets) [reply include/tiny-cuda-nn/optimizers/composite.h:20]: Do you have any clang-format style of file that you follow?

**Comment by Tom94** [comment]
Hi, thank you very much for making adjustments! I just wanted to let you know I haven't forgotten / abandoned your PR. I plan to get to it again next week.

**Comment by Tom94** [comment]
Thanks again!
